sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala

```diff
@@ -885,6 +886,12 @@ private[execution] object AggregateProcessor {
     val evaluateExpressions = mutable.Buffer.fill[Expression](ordinal)(NoOp)
     val imperatives = mutable.Buffer.empty[ImperativeAggregate]
 
+    // SPARK-14244: `SizeBasedWindowFunction`s created on driver side and serialized to executor
+    // side contains global window partition size attribute reference (`SizeBasedWindowFunction.n`),
+    // and must be rewritten using the one instantiated on executor side. Otherwise, attribute
+    // binding fails because of different expression ID.
+    functions.foreach(rewriteWindowPartitionSize)
```

Comment:
After rethinking about it, reusing driver side attribute indeed makes more sense, which is also what the first commit of this PR does.